### ISSUE TITLE
Make registry client private for v3

### ISSUE THREAD

**Issue by corhere** [distribution/distribution#4119]
https://github.com/distribution/distribution/issues/4110#issuecomment-1771170642

**Comment by milosgajdos** [comment]
I wonder if we should rip out the proxy, too. Proxy is the only piece in this repo that actually uses the client.

Proxy is also in a very dire state -- not very well maintained, swarming with races or whatnot. It needs a serious rework.

IMHO the first step could be moving the proxy [and the client] into a separate dedicated repo and mark them as experimental. Maybe.

**Comment by Jamstah** [comment]
There are lots of PRS for the proxy which implies it gets active use. It would be a shame to lose pull through support.

**Comment by Jamstah** [comment]
Although maybe if we document how to do it with a standard http proxy instead? I'm not sure what we're adding beyond that.

**Comment by Jamstah** [comment]
@wy65701436 does the harbour proxy rely on the distribution proxy?

**Comment by milosgajdos** [comment]
> There are lots of PRS for the proxy which implies it gets active use. It would be a shame to lose pull through support.

I am not saying we should kill it. I'm saying we should move it to a dedicated repo. Maybe. Probably.

It's an important piece of people's infras -- folks use it to avoid Docker pull count limits, it caches images in K8s clusters in home labs thus minimising pull ingress and roundtrips, etc.

All I'm saying is, it's in a halfbroken state at the moment where it's just not gotten enough attention for a long time. If we move it to a dedicated repo and properly focus on it in the roadmap for v4 we can make it work better then.

**Comment by Jamstah** [comment]
What do you mean by move it out? Build a second binary for proxies in another repo?

**Comment by milosgajdos** [comment]
> What do you mean by move it out? Build a second binary for proxies in another repo?

Yes. I'm not saying we must do it, I'm just saying we should consider doing that. Or maybe we could do it for v4. Dunno.

The main reason I'm thinking this may be good, other than what I mentioned above is that I did a quick code skimming regarding some of the issues @corhere raised. 

As valid as they are addressing [some of] them would require some fundamental changes that with the current abstractions leaking all over the place (including proxy and client code) would delay the v3 release even more.

Given you and I are the only people who somewhat actively **contribute** the code in this repo I dont think it's worth addressing those issues thus further delaying v3 🤷‍♂️

To summarise the options:
* keep the proxy and the client in for v3; make v3 release with them but do not address the issues raised by @corhere
* rip them out into dedicated repos, go ahead with v3 and fix the proxy/client properly (probably not client) in v4; with v3 out and these slated for v4, maybe we can get folks to focus on these more? 🤷‍♂️ 

Do you have a better suggestion?

**Comment by Jamstah** [comment]
Is there a problem with releasing v3 without removing anything?

**Comment by milosgajdos** [comment]
> Is there a problem with releasing v3 without removing anything?

There is not. I personally do not believe so. 

But the milestone has been updated with the new issues I've mentioned here:
https://github.com/distribution/distribution/milestone/22

**Comment by Jamstah** [comment]
I'd be happy with that.

The easy thing we could do is update documentation, but we already say the client is deprecated and document the limitations of the proxy.

This feels like something to have a meeting about: https://hackmd.io/TY0M7ZJ0SfO5qGC_hoJOkA

**Comment by corhere** [comment]
Another option to consider would be to make client private by moving the packages under an `internal` path. The proxy could continue to function unmodified aside from some changes to import paths, and we are afforded the flexibility to do whatever we want to the client code freed from any of the constraints of carrying a deprecated-but-public client through the entire lifecycle of the v3 module.

Shredding the project across a bunch of different repos would add a lot of overhead to development. And from a technical perspective, it wouldn't be entirely trivial to convert the proxy into an out-of-tree middleware. While it is mostly registry-middleware-shaped already, [there are a couple of hooks into the core app code](https://github.com/distribution/distribution/blob/708bc6f3e92dd35eaabc31a4a1bbf006270b9e50/registry/handlers/app.go#L82-L83) which would need to be redesigned somehow. And when it comes to out-of-tree middlewares/plugins/etc., it can be really easy to fall into the trap of circular module imports. The common imports (type definitions, enums, etc.) for both sides of the middleware API have to be moved into a separate module to break the cycle. A lot of careful thought and consideration has to go into getting it right, and we are probably not going to get it right if we rush to get it done for v3.0.0. If we also moved the proxy into an `internal` path for v3.0.0, we'd be free to refactor it into a self-contained middleware and move it out-of-tree (or not) at any time in the future, without needing to wait for v4.

**Comment by thaJeztah** [comment]
> Proxy is also in a very dire state -- not very well maintained, swarming with races or whatnot. It needs a serious rework.

I think it'd be good to discuss that separate from this. I know for sure that mirroring a registry is a much-used use-case for this code. It may not be perfect/great, but removing that functionality will probably cause quite an uproar, so not sure if we should.

> We already say people shouldn't use it. Do we even need to put in it another repo?

(https://github.com/distribution/distribution/issues/4110#issuecomment-1771170642)

Also looked a bit at this (we discussed some of this during Yesterday's Moby maintainers call). I agree that the client isn't great, and it's actually pretty (over)complicated. The main reason it got integrated into Moby (and BuildKit) is that this repository was _the_ implementation of the registry. There was no (OCI) specification; this repository _was_ the specification, and with a still evolving implementation, using the client that was designed for the implementation was the best bet. There were definitely rough edges and under-specified behavior, so using the implentation from this repository allowed us to interact with the registry (and keeping the expertise in that area to the people working on the implementation and specification).

The second reason to use this repository was because there were still parts defined here for which this repository was to be considered the canonical implementation / specification; most notably the `reference` parsing, which we now have moved separate (:tada:)

Now that the OCI specifications are in place, creating an alternative implementation should have gotten a lot more realistic, and would also address some of the package organisation in this repository (the client also having dependencies on the API _server_ implementation (gorlla/mux), which definitely is out of place for a client :sweat_smile:).

From my/our (moby) side; :+1: moving on to either a different client implementation (which could be the one from containerd) and/or (temporarily) having an in-tree fork of the code we need sounds ok to me.

That is all "in principle" though; there are still parts in this repository that are not part of the OCI spec (authentication, most notably), so as long as there's no issue in re-implementing that (and making sure it's kept in sync with any changes to those parts if they may happen), I don't have any issues.

As to actively _removing_ the client for v3, making it "internal", or moving it elsewhere;

- I think it's becoming time to start looking at doing an actual v3 release (at least an alpha or beta). I know the intent is to take the opportunity of a major (v3) release to also get rid of cruft, I also don't want that to result in "v3 (unreleased)" forever. v3 _is_ a major release, but nothing prevents us from doing a v4 following that. For sure, go modules make those slightly more disruptive (for those that still have reasons to consume packages from this repository), but should become less of an issue if such packages are either removed or moved elsewhere.
- If there is no intent to rewrite the existing client code into something more consumable, then I think For v3 I'd be good with _keeping_ the package, but marking it as deprecated. Doing so would not penalize those that _did_ move to v3 immediately, but gives a clear message that they should move to something else for the client code (and remove it or make it internal in v4)

But of course Question remains; if this repository is considered a "reference" implementation of the OCI distribution spec, should there also be an implementation for client-code _somewhere_? I could still see a use of a common implementation (but hopefully something better designed than the current one).

**Comment by corhere** [comment]
> move to v3

v2 and v3 can be imported simultaneously so nobody consuming distribution/distribution as a library will be forced to drop the v2 imports if they start importing v3 packages. **Moving the client private in v3.0.0 will penalize nobody.** It would afford us more flexibility to iterate on the private code and make it public again in the future _without having to make a v4 release_. I am advocating for this approach because it lets us get rid of the cruft _without_ getting stuck in unreleased analysis paralysis forever.

distribution/distribution/v3 is a new module with a completely blank API stability contract. We can apply the same "no is temporary, yes is forever" approach as we do for the moby Engine API by erring on the side of moving all questionable packages under internal for v3.0.0. Keeping it internal is temporary and reversible; making it public is forever (for the github.com/distribution/distribution/v3 module). The actual mechanics of moving the packages is easy enough with `gomvpkg`, and since making packages private in a module before it has stabilized affects nobody there is no downside to making more packages private than necessary.
